iina/MainWindowTouchBarSupport.swift

```diff
@@ -0,0 +1,252 @@
+  .fastForward: (NSImageNameTouchBarFastForwardTemplate, 1, "Fast forward")
+]
+
+@available(OSX 10.12.2, *)
+extension MainWindowController: NSTouchBarDelegate {
+
+  override func makeTouchBar() -> NSTouchBar? {
+    let touchBar = NSTouchBar()
+    touchBar.delegate = self
+    touchBar.customizationIdentifier = .windowBar
+    touchBar.defaultItemIdentifiers = [.playPause, .slider, .time]
+    touchBar.customizationAllowedItemIdentifiers = [.playPause, .slider, .volumeUp, .volumeDown, .rewind, .fastForward, .time, .ahead15Sec, .ahead30Sec, .back15Sec, .back30Sec, .next, .prev, .fixedSpaceLarge]
+    return touchBar
+  }
+
+  func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItemIdentifier) -> NSTouchBarItem? {
+
+    switch identifier {
+
+    case NSTouchBarItemIdentifier.playPause:
+      let item = NSCustomTouchBarItem(identifier: identifier)
+      item.view = NSButton(image: NSImage(named: NSImageNameTouchBarPauseTemplate)!, target: self, action: #selector(self.touchBarPlayBtnAction(_:)))
+      item.customizationLabel = "Play / Pause"
+      self.touchBarPlayPauseBtn = item.view as? NSButton
+      return item
+
+    case NSTouchBarItemIdentifier.slider:
+      let item = NSSliderTouchBarItem(identifier: identifier)
+      item.slider.cell = TouchBarPlaySliderCell()
+      item.slider.minValue = 0
+      item.slider.maxValue = 100
+      item.slider.target = self
+      item.slider.action = #selector(self.touchBarSliderAction(_:))
+      item.customizationLabel = "Seek"
+      self.touchBarPlaySlider = item.slider
+      return item
+
+    case NSTouchBarItemIdentifier.volumeUp,
+         NSTouchBarItemIdentifier.volumeDown:
+      guard let data = touchBarItemBinding[identifier] else { return nil }
+      return buttonTouchBarItem(withIdentifier: identifier, imageName: data.0, tag: data.1, customLabel: data.2, action: #selector(self.touchBarVolumeAction(_:)))
+
+    case NSTouchBarItemIdentifier.rewind,
+         NSTouchBarItemIdentifier.fastForward:
+      guard let data = touchBarItemBinding[identifier] else { return nil }
+      return buttonTouchBarItem(withIdentifier: identifier, imageName: data.0, tag: data.1, customLabel: data.2, action: #selector(self.touchBarRewindAction(_:)))
+
+    case NSTouchBarItemIdentifier.time:
+      let item = NSCustomTouchBarItem(identifier: identifier)
+      let label = DurationDisplayTextField(labelWithString: "00:00")
+      label.alignment = .center
+      label.mode = ud.bool(forKey: Preference.Key.showRemainingTime) ? .remaining : .current
+      self.touchBarCurrentPosLabel = label
+      item.view = label
+      item.customizationLabel = "Time Position"
+      return item
+
+    case NSTouchBarItemIdentifier.ahead15Sec,
+         NSTouchBarItemIdentifier.back15Sec,
+         NSTouchBarItemIdentifier.ahead30Sec,
+         NSTouchBarItemIdentifier.back30Sec:
+      guard let data = touchBarItemBinding[identifier] else { return nil }
```

Comment:
`guard` is designed for guarding and nil test. Like the sample in documentation, the most valuable usage of `flatMap` on `Any?` should be producing optional values though complicated logic in order to be embedded in optional chain.

```
let possibleNumber: Int? = Int("42")
let nonOverflowingSquare = possibleNumber.flatMap { x -> Int? in
    let (result, overflowed) = Int.multiplyWithOverflow(x, x)
    return overflowed ? nil : result
}
```

It's totally my personal opinion, but I still prefer `guard` for logical simplicity.